### PR TITLE
Issue #688 Add support for WsTrust 1.3

### PR DESCRIPTION
**Pull request by lovemaths**

### CODE REVIEW

**Comment by lovemaths** [comment]
@brentschmaltz changed List to ICollection

**Comment by Zoxive** [comment]
At first glance it doesn't appear to work for me. 

I tried forking and adding a test to show you it failing to parse my XML but I couldn't get tests to run locally, they fail with 

> Message: System.IO.FileLoadException : Could not load file or assembly 'Microsoft.IdentityModel.Protocols.WsFederation, Version=5.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. Strong name signature could not be verified.  The assembly may have been tampered with, or it was delay signed but not fully signed with the correct private key. (Exception from HRESULT: 0x80131045)

Anyways..

The main issue seems to be that my WResult first element is `<trust:RequestSecurityTokenResponseCollection xmlns:trust=""http://docs.oasis-open.org/ws-sx/ws-trust/200512"">`

When its expecting the first element to be RequestSecurityTokenResponse (which is the inside RequestSecurityTokenResponseCollection

File: src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMessage.cs [hunk 1]
```diff
@@ -181,13 +181,20 @@ public virtual string GetToken()
                 XmlReader xmlReader = XmlReader.Create(sr);
                 xmlReader.MoveToContent();
 
-                XmlUtil.CheckReaderOnEntry(xmlReader, WsTrustConstants.Elements.RequestSecurityTokenResponse, WsTrustConstants.Namespaces.WsTrust2005);
+                var namespaceList = new List<string>()
```
Comment (Zoxive) [inline src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMessage.cs]: Probably want to move these somewhere and make them static so we don't allocate them everytime.

File: src/Microsoft.IdentityModel.Xml/XmlUtil.cs [hunk 2]
```diff
@@ -93,6 +94,71 @@ public static void CheckReaderOnEntry(XmlReader reader, string element, string @
...
+                throw LogArgumentNullException(nameof(reader));
+
+            // IsStartElement calls reader.MoveToContent().
+            if (!reader.IsStartElement())
+                throw LogReadException(LogMessages.IDX21022, reader.NodeType);
+
+            if (namespaceList == null || namespaceList.Count == 0)
+            {
+                if (!reader.IsStartElement(element))
+                    throw LogReadException(LogMessages.IDX21024, element, reader.LocalName);
```
Comment (Zoxive) [inline src/Microsoft.IdentityModel.Xml/XmlUtil.cs]: This error doesn't seem to give the right message.

**Comment by lovemaths** [comment]
@Zoxive Can you share with me your complete WResult?

**Comment by Zoxive** [comment]
Sure I don't think there's anything sensitive in there, its all local development data.

https://gist.github.com/Zoxive/1e4d32b2d052fe06dd00198e3344b1e2

My quick googling it appears like the collection is the first element for 200512 ws-trust-1.3
http://docs.oasis-open.org/ws-sx/ws-trust/200512/ws-trust-1.3-os.pdf

**Comment by lovemaths** [comment]
@Zoxive Cool. Thank you!

File: src/Microsoft.IdentityModel.Xml/XmlUtil.cs [hunk 3]
```diff
@@ -93,6 +94,71 @@ public static void CheckReaderOnEntry(XmlReader reader, string element, string @
...
+        /// Checks if the <see cref="XmlReader"/> is pointing to an expected element.
+        /// </summary>
+        /// <param name="reader">the <see cref="XmlReader"/>to check.</param>
+        /// <param name="element">the expected element.</param>
+        /// <param name="namespaceList">the expected namespace list.</param>
+        /// <exception cref="ArgumentNullException">if <paramref name="reader"/> is null.</exception>
+        /// <exception cref="ArgumentNullException">if <paramref name="element"/> is null or empty.</exception>
+        /// <exception cref="XmlReadException">if <paramref name="reader"/> if not at a StartElement.</exception>
+        /// <exception cref="XmlReadException">if <paramref name="reader"/> if not at expected element.</exception>
+        public static void CheckReaderOnEntry(XmlReader reader, string element, ICollection<string> namespaceList)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Xml/XmlUtil.cs]: Can we make this internal? If we ever support WsTrust we would handle multiple versions in a different way. For plucking a token out of a WsFed message, it seems fine.

**Comment by brentschmaltz** [comment]
@Zoxive did you get the " Strong name signature ..." issue solved?
We added two reg key files that can be used to bypass strong name.

Here is one: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/dev/build/strongNameBypass.reg

**Comment by lovemaths** [comment]
@brentschmaltz @Zoxive Updated the PR. Tried your wresult xml, now it works.

File: src/Microsoft.IdentityModel.Protocols.WsFederation/LogMessages.cs [hunk 4]
```diff
@@ -40,6 +40,7 @@ internal static class LogMessages
         internal const string IDX10900 = "IDX10900: Building wsfederation message from query string: '{0}'.";
         internal const string IDX10901 = "IDX10901: Building wsfederation message from uri: '{0}'.";
         internal const string IDX10902 = "IDX10902: Token is not found in wresult";
+        internal const string IDX10903 = "IDX10903: More than one token are found in wresult";
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Protocols.WsFederation/LogMessages.cs]: Multiple tokens were found in the RequestSecurityTokenCollection. Only a single token is supported.

File: src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMessage.cs [hunk 5]
```diff
@@ -181,37 +182,55 @@ public virtual string GetToken()
...
-                                writer.Flush();
-                            }
-
-                            ms.Seek(0, SeekOrigin.Begin);
+                            xmlReader.Skip();
+                        }
+                        else
+                        {
+                            if (token != null)
+                                throw new WsFederationException(LogMessages.IDX10903);
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMessage.cs]: add to comments

File: test/Microsoft.IdentityModel.Protocols.WsFederation.Tests/WsFederationMessageTests.cs [hunk 6]
```diff
@@ -279,12 +185,54 @@ public static TheoryData<WsFederationMessageTheoryData> GetTokenTheoryData
...
+                            }
+                        },
+                        Token = ReferenceXml.Token_Dummy,
+                        TestId = "WsFederationMessage getToken test with WsTrust 1.4"
                     }
                 };
             }
         }
 
+#pragma warning disable CS3016 // Arrays as attribute arguments is not CLS-compliant
```
Comment (brentschmaltz) [inline test/Microsoft.IdentityModel.Protocols.WsFederation.Tests/WsFederationMessageTests.cs]: We can have a single #pragma clause set. One at the beginning of the file, closing at the end.

**Comment by lovemaths** [comment]
@brentschmaltz  updated

**Comment by brentschmaltz** [comment]
:shipit:

**Comment by Zoxive** [comment]
Trying it out right now.

**Comment by Zoxive** [comment]
Seems to work. I get as far as trying to validate the token.

Don't think its this pull requests problem but Microsoft.IdentityModel.Tokens.Saml.SamlSecurityTokenHandler fails to validate with 

> 'IDX21025: Unable to read XML. Expecting XmlReader to be at EndElement: 'Signature'. Found XmlNode 'type.name': 'Element.KeyInfo'.'

**Comment by brentschmaltz** [comment]
@Zoxive any chance you can share the token that is failing?

**Comment by Zoxive** [comment]
Ill try and find a little time this weekend to get both working/not working tokens. But as i said feels like its out of scope for this PR. Want me to make a different issue?

**Comment by lovemaths** [comment]
@Zoxive Let's open a new issue for that. Thank you.

**Comment by brentschmaltz** [comment]
@Zoxive it may be out of scope for this PR, but not for the 5.2.0 release. In fact it seems very important.

**Pull request merged**